rrdtool-1.4.7/doc/rrdupdate.1
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "RRDUPDATE 1"
.TH RRDUPDATE 1 "2009-06-02" "1.4.7" "rrdtool"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
rrdupdate \- Store a new set of values into the RRD
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBrrdtool\fR {\fBupdate\fR | \fBupdatev\fR} \fIfilename\fR
[\fB\-\-template\fR|\fB\-t\fR\ \fIds-name\fR[\fB:\fR\fIds-name\fR]...]
[\fB\-\-daemon\fR\ \fIaddress\fR] [\fB\-\-\fR]
\&\fBN\fR|\fItimestamp\fR\fB:\fR\fIvalue\fR[\fB:\fR\fIvalue\fR...]
\&\fIat-timestamp\fR\fB@\fR\fIvalue\fR[\fB:\fR\fIvalue\fR...]
[\fItimestamp\fR\fB:\fR\fIvalue\fR[\fB:\fR\fIvalue\fR...]\ ...]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBupdate\fR function feeds new data values into an \fB\s-1RRD\s0\fR. The data
is time aligned (interpolated) according to the properties of the
\&\fB\s-1RRD\s0\fR to which the data is written.
.IP "\fBupdatev\fR" 8
.IX Item "updatev"
This alternate version of \fBupdate\fR takes the same arguments and
performs the same function. The \fIv\fR stands for \fIverbose\fR, which
describes the output returned. \fBupdatev\fR returns a list of any and all
consolidated data points (CDPs) written to disk as a result of the
invocation of update. The values are indexed by timestamp (time_t),
\&\s-1RRA\s0 (consolidation function and PDPs per \s-1CDP\s0), and data source (name).
Note that depending on the arguments of the current and previous call to
update, the list may have no entries or a large number of entries.
.Sp
Since \fBupdatev\fR requires direct disk access, the \fB\-\-daemon\fR option cannot be
used with this command.
.IP "\fIfilename\fR" 8
.IX Item "filename"
The name of the \fB\s-1RRD\s0\fR you want to update.
.IP "\fB\-\-template\fR|\fB\-t\fR \fIds-name\fR[\fB:\fR\fIds-name\fR]..." 8
.IX Item "--template|-t ds-name[:ds-name]..."
By default, the \fBupdate\fR function expects its data input in the order
the data sources are defined in the \s-1RRD\s0, excluding any \s-1COMPUTE\s0 data
sources (i.e. if the third data source \fB\s-1DST\s0\fR is \s-1COMPUTE\s0, the third
input value will be mapped to the fourth data source in the \fB\s-1RRD\s0\fR and
so on). This is not very error resistant, as you might be sending the
wrong data into an \s-1RRD\s0.
.Sp
The template switch allows you to specify which data sources you are
going to update and in which order. If the data sources specified in
the template are not available in the \s-1RRD\s0 file, the update process
will abort with an error message.
.Sp
While it appears possible with the template switch to update data sources
asynchronously, \fBRRDtool\fR implicitly assigns non-COMPUTE data sources missing
from the template the \fI*UNKNOWN*\fR value.
.Sp
Do not specify a value for a \s-1COMPUTE\s0 \fB\s-1DST\s0\fR in the \fBupdate\fR
function. If this is done accidentally (and this can only be done
using the template switch), \fBRRDtool\fR will ignore the value specified
for the \s-1COMPUTE\s0 \fB\s-1DST\s0\fR.
.IP "\fB\-\-daemon\fR \fIaddress\fR" 8
.IX Item "--daemon address"
If given, \fBRRDTool\fR will try to connect to the caching daemon rrdcached
at \fIaddress\fR and will fail if the connection cannot be established. If the
connection is successfully established the values will be sent to the daemon
instead of accessing the files directly.
.Sp
For a list of accepted formats, see the \fB\-l\fR option in the rrdcached manual.
.IP "\fBN\fR|\fItimestamp\fR\fB:\fR\fIvalue\fR[\fB:\fR\fIvalue\fR...]" 8
.IX Item "N|timestamp:value[:value...]"
The data used for updating the \s-1RRD\s0 was acquired at a certain
time. This time can either be defined in seconds since 1970\-01\-01 or
by using the letter 'N', in which case the update time is set to be
the current time. Negative time values are subtracted from the current
time. An \s-1AT_STYLE\s0 \s-1TIME\s0 \s-1SPECIFICATION\s0 (see the \fIrrdfetch\fR
documentation) may also be used by delimiting the end of the time
specification with the '@' character instead of a ':'. Getting the
timing right to the second is especially important when you are
working with data-sources of type \fB\s-1COUNTER\s0\fR, \fB\s-1DERIVE\s0\fR or
\&\fB\s-1ABSOLUTE\s0\fR.
.Sp
When using negative time values, options and data have to be separated
by two dashes (\fB\-\-\fR), else the time value would be parsed as an option.
See below for an example.
.Sp
When using negative time values, options and data have to be separated
by two dashes (\fB\-\-\fR), else the time value would be parsed as an option.
See below for an example.
.Sp
The remaining elements of the argument are \s-1DS\s0 updates. The order of
this list is the same as the order the data sources were defined in
the \s-1RRA\s0. If there is no data for a certain data-source, the letter
\&\fBU\fR (e.g., N:0.1:U:1) can be specified.
.Sp
The format of the value acquired from the data source is dependent on
the data source type chosen. Normally it will be numeric, but the data
acquisition modules may impose their very own parsing of this
parameter as long as the colon (\fB:\fR) remains the data source value
separator.
.SH "ENVIRONMENT VARIABLES"
.IX Header "ENVIRONMENT VARIABLES"
The following environment variables may be used to change the behavior of
\&\f(CW\*(C`rrdtool\ update\*(C'\fR:
.IP "\fB\s-1RRDCACHED_ADDRESS\s0\fR" 4
.IX Item "RRDCACHED_ADDRESS"
If this environment variable is set it will have the same effect as specifying
the \f(CW\*(C`\-\-daemon\*(C'\fR option on the command line. If both are present, the command
line argument takes precedence.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.IP "\(bu" 4
\&\f(CW\*(C`rrdtool update demo1.rrd N:3.44:3.15:U:23\*(C'\fR
.Sp
Update the database file demo1.rrd with 3 known and one \fI*UNKNOWN*\fR
value. Use the current time as the update time.
.IP "\(bu" 4
\&\f(CW\*(C`rrdtool update demo2.rrd 887457267:U 887457521:22 887457903:2.7\*(C'\fR
.Sp
Update the database file demo2.rrd which expects data from a single
data-source, three times. First with an \fI*UNKNOWN*\fR value then with two
regular readings. The update interval seems to be around 300 seconds.
.IP "\(bu" 4
\&\f(CW\*(C`rrdtool update demo3.rrd \-\- \-5:21 N:42\*(C'\fR
.Sp
Update the database file demo3.rrd two times, using five seconds in the
past and the current time as the update times.
.IP "\(bu" 4
\&\f(CW\*(C`rrdtool update \-\-cache /var/lib/rrd/demo3.rrd N:42\*(C'\fR
.Sp
Update the file \f(CW\*(C`/var/lib/rrd/demo3.rrd\*(C'\fR with a single data source, using the
current time. If the caching daemon cannot be reached, do \fBnot\fR fall back to
direct file access.
.IP "\(bu" 4
\&\f(CW\*(C`rrdtool update \-\-daemon unix:/tmp/rrdd.sock demo4.rrd N:23\*(C'\fR
.Sp
Use the \s-1UNIX\s0 domain socket \f(CW\*(C`/tmp/rrdd.sock\*(C'\fR to contact the caching daemon. If
the caching daemon is not available, update the file \f(CW\*(C`demo4.rrd\*(C'\fR directly.
\&\fB\s-1WARNING:\s0\fR Since a relative path is specified, the following disturbing effect
may occur: If the daemon is available, the file relative to the working
directory \fBof the daemon\fR is used. If the daemon is not available, the file
relative to the current working directory of the invoking process is used.
\&\fBThis may update two different files depending on whether the daemon could be
reached or not.\fR Don't do relative paths, kids!
.SH "AUTHORS"
.IX Header "AUTHORS"
Tobias Oetiker <tobi@oetiker.ch>,
Florian Forster <octo\ at\ verplant.org>


rrdtool-1.4.7/src/rrdupdate.c
/*****************************************************************************
 * RRDtool 1.4.7  Copyright by Tobi Oetiker, 1997-2012
 *****************************************************************************
 * rrdupdate.c  Main program for the (standalone) rrdupdate utility
 *****************************************************************************
 * $Id: rrdupdate.c 2267 2012-01-24 10:08:48Z oetiker $
 *****************************************************************************/

#if defined(_WIN32) && !defined(__CYGWIN__) && !defined(__CYGWIN32__) && !defined(HAVE_CONFIG_H)
#include "../win32/config.h"
#else
#ifdef HAVE_CONFIG_H
#include "../rrd_config.h"
#endif
#endif

#include "rrd.h"

int main(
    int argc,
    char **argv)
{
    rrd_update(argc, argv);
    if (rrd_test_error()) {
        printf("RRDtool " PACKAGE_VERSION
               "  Copyright by Tobi Oetiker, 1997-2010\n\n"
               "Usage: rrdupdate filename\n"
               "\t\t\t[--template|-t ds-name:ds-name:...]\n"
               "\t\t\ttime|N:value[:value...]\n\n"
               "\t\t\tat-time@value[:value...]\n\n"
               "\t\t\t[ time:value[:value...] ..]\n\n");

        printf("ERROR: %s\n", rrd_get_error());
        rrd_clear_error();
        return 1;
    }
    return 0;
}
